Mapping to existing entity is broken when existing entity has more then 20.000 rows

### How to Reproduce
Upload the content of the mappingservicebug.zip in the projects/Leiden folder on drive. Dont forget to alter the id of the admin user. 
Go to mapping service. Map to leiden_RP
Same thing happens when first mapping to RP (which succeeds) and then mapping again to RP

### Expected behavior
Mapping succeeds, leiden_RP is updated with age categories

### Observed behavior
On 3.1 (SNAPSHOT, master)
Failed to create integrated entity.

; nested exception is org.springframework.dao.DataAccessResourceFailureException: PreparedStatementCallback; SQL [UPDATE "RP#268f9e95" SET "BBMRI_id" = ?, "rp2_id" = ?, "rp3_id" = ?, "rp4_id" = ?, "biobank" = ?, "wgs" = ?, "genotypes" = ?, "methylome" = ?, "transcriptome" = ?, "metabolome" = ?, "age" = ?, "sex" = ?, "smoking" = ?, "wbcc" = ?, "ageGroup" = ? WHERE "BBMRI_id"= ?]; ERROR: out of shared memory Hint: You might need to increase max_locks_per_transaction.; nested exception is org.postgresql.util.PSQLException: ERROR: out of shared memory Hint: You might need to increase max_locks_per_transaction.

On 3.0, 500 error with in log: 
`Caused by: org.postgresql.util.PSQLException: ERROR: out of shared memory
  Hint: You might need to increase max_locks_per_transaction.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2458) ~[postgresql-9.4.1211.jar:9.4.1211]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2158) ~[postgresql-9.4.1211.jar:9.4.1211]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:463) ~[postgresql-9.4.1211.jar:9.4.1211]
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:794) ~[postgresql-9.4.1211.jar:9.4.1211]
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1662) ~[postgresql-9.4.1211.jar:9.4.1211]
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2544) ~[c3p0-0.9.5.2.jar:0.9.5.2]
	at org.springframework.jdbc.core.JdbcTemplate$4.doInPreparedStatement(JdbcTemplate.java:966) ~[spring-jdbc-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate$4.doInPreparedStatement(JdbcTemplate.java:950) ~[spring-jdbc-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:633) ~[spring-jdbc-4.3.4.RELEASE.jar:4.3.4.RELEASE]
	... 148 common frames omitted`

--- a/molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java
+++ b/molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java
@@ -26,7 +26,6 @@ import org.springframework.transaction.annotation.Transactional;
 
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
@@ -35,7 +34,7 @@ import static com.google.api.client.util.Maps.newHashMap;
 import static java.lang.Boolean.TRUE;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
-import static java.util.stream.Collectors.*;
+import static java.util.stream.Collectors.toList;
 import static org.molgenis.data.mapper.meta.MappingProjectMetaData.NAME;
 import static org.molgenis.data.meta.model.EntityType.AttributeCopyMode.DEEP_COPY_ATTRS;
 import static org.molgenis.data.support.EntityTypeUtils.hasSelfReferences;
@@ -357,33 +356,23 @@ public class MappingServiceImpl implements MappingService
 	private void processBatch(EntityMapping sourceMapping, Repository<Entity> targetRepo, Progress progress,
 			AtomicLong counter, boolean canAdd, List<Entity> entities)
 	{
-		Stream<Entity> mappedEntities = mapEntities(sourceMapping, targetRepo.getEntityType(), entities);
+		List<Entity> mappedEntities = mapEntities(sourceMapping, targetRepo.getEntityType(), entities);
 		if (canAdd)
 		{
-			targetRepo.add(mappedEntities);
+			targetRepo.add(mappedEntities.stream());
 		}
 		else
 		{
-			upsertBatch(targetRepo, mappedEntities.collect(toList()));
+			targetRepo.upsertBatch(mappedEntities);
 		}
 		progress.increment(1);
 		counter.addAndGet(entities.size());
 	}
 
-	private static void upsertBatch(Repository<Entity> targetRepo, List<Entity> entities)
+	private List<Entity> mapEntities(EntityMapping sourceMapping, EntityType targetMetaData, List<Entity> entities)
 	{
-		Set<Object> ids = entities.stream().map(Entity::getIdValue).collect(toSet());
-		Fetch idFetch = new Fetch().field(targetRepo.getEntityType().getIdAttribute().getName());
-		Set<Object> existingIDs = targetRepo.findAll(ids.stream(), idFetch).collect(toSet());
-		Map<Boolean, List<Entity>> partitioned = entities.stream()
-				.collect(partitioningBy(entity -> existingIDs.contains(entity.getIdValue())));
-		targetRepo.add(partitioned.get(Boolean.FALSE).stream());
-		targetRepo.update(partitioned.get(Boolean.TRUE).stream());
-	}
-
-	private Stream<Entity> mapEntities(EntityMapping sourceMapping, EntityType targetMetaData, List<Entity> entities)
-	{
-		return entities.stream().map(sourceEntity -> applyMappingToEntity(sourceMapping, sourceEntity, targetMetaData));
+		return entities.stream().map(sourceEntity -> applyMappingToEntity(sourceMapping, sourceEntity, targetMetaData))
+				.collect(toList());
 	}
 
 	private Entity applyMappingToEntity(EntityMapping sourceMapping, Entity sourceEntity, EntityType targetMetaData)

--- a/molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java
+++ b/molgenis-data-mapper/src/main/java/org/molgenis/data/mapper/service/impl/MappingServiceImpl.java
@@ -26,6 +26,7 @@ import org.springframework.transaction.annotation.Transactional;
 
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
@@ -34,7 +35,7 @@ import static com.google.api.client.util.Maps.newHashMap;
 import static java.lang.Boolean.TRUE;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
-import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.*;
 import static org.molgenis.data.mapper.meta.MappingProjectMetaData.NAME;
 import static org.molgenis.data.meta.model.EntityType.AttributeCopyMode.DEEP_COPY_ATTRS;
 import static org.molgenis.data.support.EntityTypeUtils.hasSelfReferences;
@@ -369,20 +370,15 @@ public class MappingServiceImpl implements MappingService
 		counter.addAndGet(entities.size());
 	}
 
-	private static void upsertBatch(Repository<Entity> targetRepo, List<Entity> mappedEntities)
+	private static void upsertBatch(Repository<Entity> targetRepo, List<Entity> entities)
 	{
-		mappedEntities.forEach(mappedEntity ->
-		{
-			// FIXME adding/updating row-by-row is a performance bottleneck, this code could do streaming upsert
-			if (targetRepo.findOneById(mappedEntity.getIdValue()) == null)
-			{
-				targetRepo.add(mappedEntity);
-			}
-			else
-			{
-				targetRepo.update(mappedEntity);
-			}
-		});
+		Set<Object> ids = entities.stream().map(Entity::getIdValue).collect(toSet());
+		Fetch idFetch = new Fetch().field(targetRepo.getEntityType().getIdAttribute().getName());
+		Set<Object> existingIDs = targetRepo.findAll(ids.stream(), idFetch).collect(toSet());
+		Map<Boolean, List<Entity>> partitioned = entities.stream()
+				.collect(partitioningBy(entity -> existingIDs.contains(entity.getIdValue())));
+		targetRepo.add(partitioned.get(Boolean.FALSE).stream());
+		targetRepo.update(partitioned.get(Boolean.TRUE).stream());
 	}
 
 	private Stream<Entity> mapEntities(EntityMapping sourceMapping, EntityType targetMetaData, List<Entity> entities)

--- a/molgenis-data-mapper/src/test/java/org/molgenis/data/mapper/service/impl/MappingServiceImplTest.java
+++ b/molgenis-data-mapper/src/test/java/org/molgenis/data/mapper/service/impl/MappingServiceImplTest.java
@@ -385,9 +385,6 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		when(dataService.hasRepository(entityTypeId)).thenReturn(true);
 		when(dataService.getRepository(entityTypeId)).thenReturn(updateEntityRepo);
 
-		when(updateEntityRepo.count()).thenReturn(4L);
-		when(updateEntityRepo.findOneById(any())).thenReturn(mock(Entity.class));
-
 		when(updateEntityRepo.getEntityType()).thenReturn(targetMeta);
 		when(metaDataService.createRepository(argThat(new ArgumentMatcher<EntityType>()
 		{
@@ -402,6 +399,9 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		List<Entity> sourceGeneEntities = newArrayList();
 		List<Entity> expectedEntities = newArrayList();
 		createEntities(targetMeta, sourceGeneEntities, expectedEntities);
+
+		when(updateEntityRepo.count()).thenReturn(4L);
+		when(updateEntityRepo.findAll(any(), any())).thenReturn(expectedEntities.stream());
 
 		doAnswer(invocationOnMock ->
 		{
@@ -421,7 +421,7 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 						progress), 4);
 
 		verify(geneRepo).forEachBatched(any(Consumer.class), any(Integer.class));
-		verify(updateEntityRepo, times(4)).update(any(Entity.class));
+		verify(updateEntityRepo).update(any(Stream.class));
 
 		verify(progress).status("Applying mappings to repository [HopEntity]");
 		verify(progress).status("Mapping source [Genes]...");
@@ -484,6 +484,9 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		EntityType targetEntityType = mock(EntityType.class);
 		when(targetEntityType.getId()).thenReturn("targetEntityType");
 		when(targetEntityType.getAtomicAttributes()).thenReturn(newArrayList());
+		Attribute targetID = mock(Attribute.class);
+		when(targetID.getName()).thenReturn("targetID");
+		when(targetEntityType.getIdAttribute()).thenReturn(targetID);
 		when(targetRepo.getEntityType()).thenReturn(targetEntityType);
 
 		List<Entity> batch = newArrayList(mock(Entity.class), mock(Entity.class));
@@ -492,19 +495,18 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		{
 			Consumer<List<Entity>> consumer = (Consumer<List<Entity>>) invocationOnMock
 					.getArgumentAt(0, Consumer.class);
-
-			when(targetRepo.findOneById(any(String.class))).thenReturn(mock(Entity.class));
+			when(targetRepo.findAll(any(Stream.class), any(Fetch.class))).thenReturn(batch.stream());
 			consumer.accept(batch);
 
-			when(targetRepo.findOneById(any(String.class))).thenReturn(null);
+			when(targetRepo.findAll(any(Stream.class), any(Fetch.class))).thenReturn(Stream.empty());
 			consumer.accept(batch);
 			return null;
 		}).when(sourceRepo).forEachBatched(any(Consumer.class), eq(MAPPING_BATCH_SIZE));
 
 		mappingService.applyMappingToRepo(sourceMapping, targetRepo, progress);
 
-		verify(targetRepo, times(2)).add(any(Entity.class));
-		verify(targetRepo, times(2)).update(any(Entity.class));
+		verify(targetRepo, times(2)).add(any(Stream.class));
+		verify(targetRepo, times(2)).update(any(Stream.class));
 		verify(progress, times(2)).increment(1);
 		verify(progress).status("Mapping source [sourceMappingLabel]...");
 		verify(progress).status("Mapped 4 [sourceMappingLabel] entities.");

--- a/molgenis-data-mapper/src/test/java/org/molgenis/data/mapper/service/impl/MappingServiceImplTest.java
+++ b/molgenis-data-mapper/src/test/java/org/molgenis/data/mapper/service/impl/MappingServiceImplTest.java
@@ -2,6 +2,7 @@ package org.molgenis.data.mapper.service.impl;
 
 import org.mockito.ArgumentCaptor;
 import org.mockito.ArgumentMatcher;
+import org.mockito.Captor;
 import org.mockito.Mock;
 import org.molgenis.auth.User;
 import org.molgenis.auth.UserFactory;
@@ -23,6 +24,7 @@ import org.molgenis.data.meta.model.Package;
 import org.molgenis.data.support.DynamicEntity;
 import org.molgenis.js.magma.JsMagmaScriptEvaluator;
 import org.molgenis.security.permission.PermissionSystemService;
+import org.molgenis.util.EntityUtils;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
@@ -98,6 +100,9 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 
 	@Mock
 	private Progress progress;
+
+	@Captor
+	ArgumentCaptor<List<Entity>> batchCaptor;
 
 	private MetaDataService metaDataService;
 
@@ -421,7 +426,9 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 						progress), 4);
 
 		verify(geneRepo).forEachBatched(any(Consumer.class), any(Integer.class));
-		verify(updateEntityRepo).update(any(Stream.class));
+
+		verify(updateEntityRepo).upsertBatch(batchCaptor.capture());
+		assertTrue(EntityUtils.equalsEntities(batchCaptor.getValue(), expectedEntities));
 
 		verify(progress).status("Applying mappings to repository [HopEntity]");
 		verify(progress).status("Mapping source [Genes]...");
@@ -446,6 +453,14 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		when(sourceRepo.getEntityType()).thenReturn(sourceEntityType);
 		when(dataService.getRepository("sourceMappingID")).thenReturn(sourceRepo);
 		when(targetRepo.count()).thenReturn(0L);
+
+		EntityType targetEntityType = mock(EntityType.class);
+		when(targetEntityType.getId()).thenReturn("targetEntityType");
+		when(targetEntityType.getAtomicAttributes()).thenReturn(newArrayList());
+		Attribute targetID = mock(Attribute.class);
+		when(targetID.getName()).thenReturn("targetID");
+		when(targetEntityType.getIdAttribute()).thenReturn(targetID);
+		when(targetRepo.getEntityType()).thenReturn(targetEntityType);
 
 		List<Entity> batch = newArrayList(mock(Entity.class));
 		doAnswer(invocationOnMock ->
@@ -495,18 +510,14 @@ public class MappingServiceImplTest extends AbstractMolgenisSpringTest
 		{
 			Consumer<List<Entity>> consumer = (Consumer<List<Entity>>) invocationOnMock
 					.getArgumentAt(0, Consumer.class);
-			when(targetRepo.findAll(any(Stream.class), any(Fetch.class))).thenReturn(batch.stream());
 			consumer.accept(batch);
-
-			when(targetRepo.findAll(any(Stream.class), any(Fetch.class))).thenReturn(Stream.empty());
 			consumer.accept(batch);
 			return null;
 		}).when(sourceRepo).forEachBatched(any(Consumer.class), eq(MAPPING_BATCH_SIZE));
 
 		mappingService.applyMappingToRepo(sourceMapping, targetRepo, progress);
 
-		verify(targetRepo, times(2)).add(any(Stream.class));
-		verify(targetRepo, times(2)).update(any(Stream.class));
+		verify(targetRepo, times(2)).upsertBatch(any(List.class));
 		verify(progress, times(2)).increment(1);
 		verify(progress).status("Mapping source [sourceMappingLabel]...");
 		verify(progress).status("Mapped 4 [sourceMappingLabel] entities.");

--- a/molgenis-data/src/main/java/org/molgenis/data/Repository.java
+++ b/molgenis-data/src/main/java/org/molgenis/data/Repository.java
@@ -31,7 +31,7 @@ public interface Repository<E extends Entity> extends Iterable<E>, Closeable
 	default Set<Object> getExistingIDs(Set<Object> ids)
 	{
 		Fetch idFetch = new Fetch().field(getEntityType().getIdAttribute().getName());
-		return findAll(ids.stream(), idFetch).collect(toSet());
+		return findAll(ids.stream(), idFetch).map(Entity::getIdValue).collect(toSet());
 	}
 
 	/**

--- a/molgenis-data/src/main/java/org/molgenis/data/Repository.java
+++ b/molgenis-data/src/main/java/org/molgenis/data/Repository.java
@@ -7,15 +7,48 @@ import org.molgenis.data.meta.model.EntityType;
 
 import java.io.Closeable;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
+
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+import static java.util.stream.Collectors.partitioningBy;
+import static java.util.stream.Collectors.toSet;
 
 /**
  * Repository gives access to a collection of Entity. Synonyms: EntityReader, EntitySource, EntityCollection
  */
 public interface Repository<E extends Entity> extends Iterable<E>, Closeable
 {
+	/**
+	 * Checks if IDs are present in this {@link Repository}.
+	 *
+	 * @param ids the set of ID values to check
+	 * @return the {@link Set} of IDs that are present in the repository
+	 */
+	default Set<Object> getExistingIDs(Set<Object> ids)
+	{
+		Fetch idFetch = new Fetch().field(getEntityType().getIdAttribute().getName());
+		return findAll(ids.stream(), idFetch).collect(toSet());
+	}
+
+	/**
+	 * Upserts a batch of entities into this repository.
+	 * Entities that are already present are updated, new entities are added.
+	 *
+	 * @param entities List of Entities to upsert
+	 */
+	default void upsertBatch(List<E> entities)
+	{
+		Set<Object> existingIDs = getExistingIDs(entities.stream().map(Entity::getIdValue).collect(toSet()));
+		Map<Boolean, List<E>> partitioned = entities.stream()
+				.collect(partitioningBy(entity -> existingIDs.contains(entity.getIdValue())));
+		add(partitioned.get(FALSE).stream());
+		update(partitioned.get(TRUE).stream());
+	}
+
 	/**
 	 * Executes a function for each batch of entities.
 	 *

--- a/molgenis-data/src/main/java/org/molgenis/util/EntityUtils.java
+++ b/molgenis-data/src/main/java/org/molgenis/util/EntityUtils.java
@@ -318,6 +318,26 @@ public class EntityUtils
 	}
 
 	/**
+	 * Returns true if an Iterable equals another Iterable.
+	 *
+	 * @param entityIterable
+	 * @param otherEntityIterable
+	 * @return
+	 */
+	public static boolean equalsEntities(Iterable<Entity> entityIterable, Iterable<Entity> otherEntityIterable)
+	{
+		List<Entity> attrs = newArrayList(entityIterable);
+		List<Entity> otherAttrs = newArrayList(otherEntityIterable);
+
+		if (attrs.size() != otherAttrs.size()) return false;
+		for (int i = 0; i < attrs.size(); ++i)
+		{
+			if (!equals(attrs.get(i), otherAttrs.get(i))) return false;
+		}
+		return true;
+	}
+
+	/**
 	 * Returns true if a Tag equals another Tag.
 	 *
 	 * @param tag

--- a/molgenis-data/src/test/java/org/molgenis/data/support/AbstractRepositoryTest.java
+++ b/molgenis-data/src/test/java/org/molgenis/data/support/AbstractRepositoryTest.java
@@ -1,7 +1,6 @@
 package org.molgenis.data.support;
 
-import org.mockito.Matchers;
-import org.mockito.Mockito;
+import org.mockito.*;
 import org.molgenis.data.Entity;
 import org.molgenis.data.Fetch;
 import org.molgenis.data.Query;
@@ -12,15 +11,11 @@ import org.testng.annotations.BeforeMethod;
 import org.testng.annotations.BeforeTest;
 import org.testng.annotations.Test;
 
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.Set;
+import java.util.*;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
 import static org.testng.Assert.assertEquals;
 
 public class AbstractRepositoryTest
@@ -28,9 +23,19 @@ public class AbstractRepositoryTest
 	private AbstractRepository abstractRepository;
 	private EntityType entityType;
 
+	@Captor
+	private ArgumentCaptor<Stream<Entity>> addStreamCaptor;
+
+	@Captor
+	private ArgumentCaptor<Stream<Entity>> updateStreamCaptor;
+
+	@Captor
+	private ArgumentCaptor<Stream<Object>> objectStreamCaptor;
+
 	@BeforeTest
 	public void beforeTest()
 	{
+		MockitoAnnotations.initMocks(this);
 		Attribute idAttr = when(mock(Attribute.class).getName()).thenReturn("id").getMock();
 		entityType = when(mock(EntityType.class).getId()).thenReturn("entity").getMock();
 		when(entityType.getIdAttribute()).thenReturn(idAttr);
@@ -96,7 +101,7 @@ public class AbstractRepositoryTest
 		Entity entity1 = when(mock(Entity.class).getIdValue()).thenReturn(id1).getMock();
 		Stream<Object> entityIds = Stream.of(id0, id1);
 
-		Mockito.doReturn(Stream.of(entity0, entity1)).when(abstractRepository).findAll(Matchers.any(Query.class));
+		doReturn(Stream.of(entity0, entity1)).when(abstractRepository).findAll(Matchers.any(Query.class));
 
 		Stream<Entity> expectedEntities = abstractRepository.findAll(entityIds);
 		assertEquals(expectedEntities.collect(Collectors.toList()), Arrays.asList(entity0, entity1));
@@ -113,11 +118,32 @@ public class AbstractRepositoryTest
 		Entity entity1 = when(mock(Entity.class).getIdValue()).thenReturn(id1).getMock();
 		Stream<Object> entityIds = Stream.of(id0, id1);
 
-		Mockito.doReturn(Stream.of(entity0, entity1)).when(abstractRepository).findAll(Matchers.any(Query.class));
+		doReturn(Stream.of(entity0, entity1)).when(abstractRepository).findAll(Matchers.any(Query.class));
 
 		Stream<Entity> expectedEntities = abstractRepository.findAll(entityIds, fetch);
 		assertEquals(expectedEntities.collect(Collectors.toList()), Arrays.asList(entity0, entity1));
 	}
 
 	//	// Note: streamFetch cannot be tested because mocking default methods is not supported by Mockito
+
+	@Test
+	public void testUpsertBatch()
+	{
+		Object id0 = "id0";
+		Object id1 = "id1";
+		Entity entity0 = when(mock(Entity.class).getIdValue()).thenReturn(id0).getMock();
+		Entity entity1 = when(mock(Entity.class).getIdValue()).thenReturn(id1).getMock();
+		List<Entity> batch = Arrays.asList(entity0, entity1);
+
+		doReturn(Stream.of(entity0)).when(abstractRepository).findAll(objectStreamCaptor.capture(), any(Fetch.class));
+		doReturn(1).when(abstractRepository).add(addStreamCaptor.capture());
+		doNothing().when(abstractRepository).update(updateStreamCaptor.capture());
+
+		abstractRepository.upsertBatch(batch);
+
+		assertEquals(addStreamCaptor.getValue().collect(Collectors.toList()), Collections.singletonList(entity1),
+				"New entity should get added.");
+		assertEquals(updateStreamCaptor.getValue().collect(Collectors.toList()), Collections.singletonList(entity0),
+				"Existing entity should get updated.");
+	}
 }

